Kill running PR builds when a new build is triggered for the same PR

mbaijal opened the pull request:
@nswamy @gautamkmr @sandeep-krishnamurthy 

This is Part 1 of a series of changes to the **current MXNet PR build Process**. 

**This change might lead to some issues on Apache Jenkins, specially with requiring some script approvals as indicated[ here](https://wiki.jenkins.io/display/JENKINS/Script+Security+Plugin)**

The code for this method has been picked as is from [this stackoverflow post](https://stackoverflow.com/questions/40760716/jenkins-abort-running-build-if-new-one-is-started) 

**Expected Behavior** - 
1. I have a added a new stage to our builds called 'Purge' (other name suggestions are welcome)
2. This stage runs on a 'mxnetlinux' slave at the moment. This means it joins the queue waiting for an idle slave before it can kill an existing PR build. If we want to avoid this, it can be changed to be scheduled to the master. 
3. The method called 'abortPreviousRunningBuilds' simply scans all build numbers for the current PR and checks their status. If it is not the current build but is still running, the status of the build is changed to Aborted. 

**Testing Done** - 
I have reproduced a multibranch pipeline job on our internal jenkins similar to apache. Tests have only been done on a feature branch and PRs. The master branch should work similarly. I have tested that -
1. When a new branch is created, the build goes through fine. 
2. If a new build is triggered on a branch (say using 'Build Now'), the purge stage works as expected. (This behavior might need to be changed specially for the master branch)
3. When a new PR is created, the build works fine
4. Any updates to the PR terminates the old build and a new one is created. 

![screen shot 2017-09-25 at 3 53 11 pm](https://user-images.githubusercontent.com/30911248/30834833-bb39756a-a209-11e7-98c1-a613d95b5fd6.png)
 
![screen shot 2017-09-25 at 3 52 16 pm](https://user-images.githubusercontent.com/30911248/30834847-d6b52a00-a209-11e7-8326-538e346a47e7.png)

nswamy commented on Jenkinsfile:
```diff
@@ -11,14 +11,36 @@ max_time = 120
+      )
+      println("Aborted previous running build #${build.number}")
+    } else {
+      println("Build is not running or is current build, not aborting - #${build.number}")
```
please remove the else part in your next commit after you have verified that everything works. no need to print 'not running' every time.

mbaijal replied:
Yes, I plan to do that. If there are too many builds, this spams the logs a lot. 
Thanks.

nswamy commented on Jenkinsfile:
```diff
@@ -116,6 +138,12 @@ def python3_gpu_ut(docker_type) {
 try {
+    stage("Purge") {
+        node('mxnetlinux') {
+            abortPreviousRunningBuilds()
```
I think this should be combined in the first stage. I don't see value in going back to the queue(i.e., to get an executor for the next stage) after that this stage.

mbaijal replied:
Sure, I can make it a part of Sanity Test since we plan to have that stage whether it is a full build or smoke test. 
@gautamkmr @sandeep-krishnamurthy any opinions?

gautamkmr replied:
LGTM.

mbaijal commented:
I merged the method call into sanity test and tested on internal Jenkins 
![screen shot 2017-09-25 at 5 54 39 pm](https://user-images.githubusercontent.com/30911248/30837578-c0e32ab8-a21a-11e7-8552-802601f32b62.png)

mbaijal commented:
Can we merge this PR? That is the only way to test it since the changes have been made to the Jenkinsfile. The [PR build ](https://builds.apache.org/blue/organizations/jenkins/incubator-mxnet/detail/PR-8034/2/pipeline)has failed in unit tests so this change has not caused it

nswamy commented on Jenkinsfile:
```diff
@@ -11,14 +11,36 @@ max_time = 120
+  hi.getItem(pname).getItem(env.JOB_BASE_NAME).getBuilds().each{ build ->
+    def exec = build.getExecutor()
+
+    if (build.number != currentBuild.number && exec != null) {
```
I have two questions here:

1. Does this apply to merge builds as well? 

2. What happens let's say you have 3 builds for the same PR, build #1, #2 and #3 and let's say the order in which the executor is acquired is #1, #3, #2. When #2 runs after #3. This code will kill #3 and start #2, it should only kill jobs whose build numbers are lower?

mbaijal replied:
1. I believe it does. Merge builds run on master, correct? I should definitely add a condition on the function call saying if 'branch_name is not master'. I wanted to test what happens on apache before adding that

2. Do the PR builds not get executors in FCFS order? In that case wouldnt build #2 always get an executo rbefore build #3? (However, I can think of scenarios where the sanity test of build #3 runs befiore build #2 and adding this condition seems like a good idea)

nswamy commented on Jenkinsfile:
```diff
@@ -11,14 +11,36 @@ max_time = 120
+//Method to Kill PR Builds that are running currently when a new build is triggered.
+def abortPreviousRunningBuilds() {
+  def hi = Hudson.instance
+  def pname = env.JOB_NAME.split('/')[0]
```
nit: better variable name, what does `pname` standfor? similary the variable `hi`

mbaijal commented:
ReviewComments: 
1. Does not abort on branch master; 
2. Added Comments
3. checks for build# < currentBuild; 
4. removed else part of print; 
5. Changed variable names

Tests: 
1. I have not tested the "master" condition on the internal setup
2. I have tested by triggering many builds almost at the same time that no build with a build number greater than the current build is terminated.

mbaijal commented on Jenkinsfile:
```diff
@@ -119,6 +143,9 @@ try {
     stage("Sanity Check") {
       timeout(time: max_time, unit: 'MINUTES') {
         node('mxnetlinux') {
+          if (env.BRANCH_NAME != "master") {
```
We can also change this condition to run this only on PR builds. Let me know if that is preferable.

nswamy replied:
I don't know what is the right answer here, please check with others on slack. We can run this now on the PR builds and then remove this condition later.
